include/faabric/util/snapshot.h

```diff
@@ -74,12 +76,45 @@ class SnapshotData
 {
   public:
     size_t size = 0;
-    uint8_t* data = nullptr;
-    int fd = 0;
+    size_t maxSize = 0;
 
     SnapshotData() = default;
 
-    std::vector<SnapshotDiff> getDirtyPages();
+    explicit SnapshotData(size_t sizeIn);
+
+    explicit SnapshotData(std::vector<uint8_t> dataIn);
+
+    SnapshotData(size_t sizeIn, size_t maxSizeIn);
+
+    SnapshotData(uint8_t* dataIn, size_t sizeIn);
+
+    SnapshotData(uint8_t* dataIn, size_t sizeIn, size_t maxSizeIn);
+
+    SnapshotData(const SnapshotData&) = delete;
+
+    SnapshotData& operator=(const SnapshotData&) = delete;
+
+    ~SnapshotData();
+
+    bool isRestorable();
+
+    void makeRestorable(const std::string& fdLabel);
+
+    void setSnapshotSize(size_t newSize);
+
+    void copyInData(std::vector<uint8_t> buffer, uint32_t offset = 0);
```

Comment:
Passing `std::vector` by-value is almost always wrong and results in extra copies, `const std::vector<uint8_t>&` makes much more sense. But also this is the perfect spot to use `std::span<uint8_t>` as it's a pointer+size pair, can be constructed easily from pointers with sizes or std containers, including sub-buffers like: `span(vector).subspan(offset)` - this doesn't do any copies, just some pointer arithmetic

Reply:
Ah yes good spot and thanks for the tip re. `span`. This PR is still very much a WIP and I'm in the middle of hacking through the refactor, so will be going through and cleaning up stuff like this at the end. I'll request a review once I'm happy with the finished version.